### PR TITLE
feat(events): Add internal event id to event metadata

### PR DESCRIPTION
**Pull request by ArunPiduguDD**
### Overview

Adds an internal event id to all events that pass through Vector. This internal id will be a UUID generated on event creation, and can be used to uniquely identify an event as it traverses through different components in a pipeline.

### Testing

Added unit test

Tested using `demo_logs` source type

### CODE REVIEW

File: lib/vector-core/proto/event.proto
Comment (pront) [inline lib/vector-core/proto/event.proto:96]: Nit: we could call `source_trace_id`

File: lib/vector-core/Cargo.toml
Comment (jorgehermo9) [inline lib/vector-core/Cargo.toml:68]: How about v7? I think it may be useful to use the time-sortable property of those

And more: https://uuid7.com/

Comment (bruceg) [reply lib/vector-core/Cargo.toml:68]: Good call, that could really be useful for relatively free.

Comment (bruceg) [reply lib/vector-core/Cargo.toml:68]: Also, I see other parts of Vector are using this crate. Could you move it to a workspace dep?

File: lib/vector-core/proto/event.proto
Comment (bruceg) [reply lib/vector-core/proto/event.proto:96]: I'd say `source_event_id`, but agree on not `internal_`

File: lib/vector-core/src/event/metadata.rs
Comment (bruceg) [inline lib/vector-core/src/event/metadata.rs:224]: I believe the `Uuid` type is `Copy`. Would it make sense to return a copy instead of a reference here?

Also, I think you could just return `&self.internal_event_id`. The `.as_ref()` is only necessary above because we're turning an `&Option<T>` into `Option<&T>`.

**Review by bruceg — COMMENTED** [review]
It would be interesting to see examples of how this could be used, and I agree with the recommendation to consider V7 UUIDs, but otherwise this makes sense to me.

File: lib/vector-core/src/event/metadata.rs
Comment (ArunPiduguDD) [reply lib/vector-core/src/event/metadata.rs:224]: I think b/c this is lightweight/won't be used often we can just return a copy

Comment (jorgehermo9) [reply lib/vector-core/src/event/metadata.rs:224]: Note that even if it was used a lot, it would be cheap, as we can consider `Copy` structs to be "free" to copy

**Comment by jorgehermo9** [comment]
Some test fail due to the fact that EventMetadata::Default() is now not deterministic (as uuid are random). Don't know what to do to fix that without involving much changes 🤔

**Comment by ArunPiduguDD** [comment]
> Some test fail due to the fact that EventMetadata::Default() is now not deterministic (as uuid are random). Don't know what to do to fix that without involving much changes 🤔
> 
> Maybe a manual impl of `PartialEq` for that struct, so that the uuid field is ignored in the comparison...

@jorgehermo9 ended up using the derivative crate to ignore PartialEq (saw this crate is used in other parts of the vector codebase)

**Comment by jorgehermo9** [comment]
yep, thats sounds like what I would do, lgtm

**Review by bruceg — APPROVED** [review]

**Comment by jorgehermo9** [comment]
I think something went wrong with the merge and broke the diff... how did you do the merge with master?

**Comment by ArunPiduguDD** [comment]
> I think something went wrong with the merge and broke the diff... how did you do the merge with master?

Strange, thought I just rebased on master, let me double check

**Comment by jorgehermo9** [comment]
I think the diff is broken again @ArunPiduguDD. Maybe should be better to fix the diff of this PR before enabling auto-merge of it?

**Comment by jorgehermo9** [comment]
I see that a *lot* of files were included in https://github.com/vectordotdev/vector/pull/21074/commits/a30e41792203f456f42142def56a1098fb0e9aad. Maybe should this PR be reviewed and approved again before enabling auto-merge? @bruceg

**Comment by ArunPiduguDD** [comment]
> I see that a _lot_ of files were included in [a30e417](https://github.com/vectordotdev/vector/commit/a30e41792203f456f42142def56a1098fb0e9aad). Maybe should this PR be reviewed and approved again before including it in the merge queue? @bruceg

Had a discussion with @bruceg and @jszwedko offline about regenerating these test fixtures, can wait for another stamp of approval before merging

**Comment by jszwedko** [comment]
Small nit: we generally like to discourage force-pushes to Vector PRs once reviews have started as it makes the PR history more difficult to understand. Instead of rebasing you can merge in `master` as needed. The PR is ultimately squashed into a single commit when merging so you don't have to worry about maintaining a clean history in the PR itself.

I'm realizing this isn't mentioned in https://github.com/vectordotdev/vector/blob/master/CONTRIBUTING.md. I'll add a note there.

File: lib/vector-core/src/event/proto.rs
Comment (bruceg) [inline lib/vector-core/src/event/proto.rs]: I'm going to assume this is an editor setting, but these (and other files) ended up being reordered differently than we customarily have them.

File: lib/vector-core/src/event/metadata.rs
Comment (bruceg) [inline lib/vector-core/src/event/metadata.rs]: Ditto this reordering. Simply leaving a blank line between the groups would have resolved it.

**Review by bruceg — CHANGES_REQUESTED** [review]
The existing code LGTM. This will need to have the new `pre-v41` added to the test suite in `lib/codecs/tests/native.rs` however.

File: lib/vector-core/src/event/proto.rs
Comment (ArunPiduguDD) [reply lib/vector-core/src/event/proto.rs]: Ah I think auto-importing `Uuid` then running `cargo fmt` does this re-ordering, will keep an eye out for this in the future

File: lib/codecs/tests/native.rs
Comment (bruceg) [inline lib/codecs/tests/native.rs:28]: I think we also need tests of `roundtrip_fixtures` and `decoding_matches`.

Comment (ArunPiduguDD) [reply lib/codecs/tests/native.rs:28]: It looks like the `decoding_matches` tests are broken due to https://github.com/vectordotdev/vector/issues/18570. Will add the `ignore` flag for the `pre-v41` test as well

**Review by bruceg — APPROVED** [review]

**Review by ArunPiduguDD — COMMENTED** [review]
/ci-run-regression

**Review by ArunPiduguDD — COMMENTED** [review]
/ci-run-regression

**Review by jszwedko — COMMENTED** [review]
/ci-run-regression

**Pull request merged**
